ScoreSaber/Core/Services/LeaderboardService.cs

```diff
@@ -1,109 +1,79 @@
-﻿using Newtonsoft.Json;
+﻿#nullable enable
+using Newtonsoft.Json;
 using System.Threading.Tasks;
 using ScoreSaber.Core.Data.Wrappers;
 using ScoreSaber.Core.Data.Models;
 using System;
 using System.Linq;
 using ScoreSaber.UI.Leaderboard;
 using System.Collections.Generic;
+using ScoreSaber.Core.Http;
 
 namespace ScoreSaber.Core.Services {
     internal class LeaderboardService {
 
-        public LeaderboardMap currentLoadedLeaderboard = null;
-        public CustomLevelLoader customLevelLoader = null;
-        public BeatmapDataLoader beatmapDataLoader = null;
+        private readonly ScoreSaberHttpClient client;
+        public LeaderboardMap? currentLoadedLeaderboard;
+        public CustomLevelLoader? customLevelLoader = null;
+        public BeatmapDataLoader? beatmapDataLoader = null;
 
-        public LeaderboardService() {
+        public LeaderboardService(ScoreSaberHttpClient client) {
+            this.client = client;
             Plugin.Log.Debug("LeaderboardService Setup");
         }
 
-        public async Task<LeaderboardMap> GetLeaderboardData(int maxMultipliedScore, BeatmapLevel beatmapLevel, BeatmapKey beatmapKey, ScoreSaber.UI.Leaderboard.ScoreSaberLeaderboardViewController.ScoreSaberScoresScope scope, int page, PlayerSpecificSettings playerSpecificSettings) {
-
-            string leaderboardUrl = GetLeaderboardUrl(beatmapKey, scope, page);
-            if (leaderboardUrl == null) {
-                currentLoadedLeaderboard = null;
-                return null;
-            }
-            string leaderboardRawData = await Plugin.HttpInstance.GetAsync(leaderboardUrl);
-            Leaderboard leaderboardData = JsonConvert.DeserializeObject<Leaderboard>(leaderboardRawData);
-
+        public async Task<LeaderboardMap> GetLeaderboardData(
+            int maxMultipliedScore,
+            BeatmapLevel beatmapLevel,
+            BeatmapKey beatmapKey,
+            ScoreSaberLeaderboardViewController.ScoreSaberScoresScope scope,
+            int page) {
+            var leaderboardData = await client.GetAsync<Leaderboard>(new Http.Endpoints.API.LeaderboardRequest(
+                leaderboardId: beatmapLevel.hasPrecalculatedData
+                    ? "ost_" + beatmapLevel.levelID
+                    : beatmapKey.levelId.Split('_')[2],
```

Comment:
This code to convert between levelId and our hash occurs so often now (with differing logic sometimes even, see rich presence) and probably should be moved to some util class.

Reply:
made it apart of beatmaputils